### ISSUE TITLE
Autofac Open Generic Decorator with multiple interface causes circular dependency resolving

### ISSUE THREAD

**Issue by H-Ahmadi** [autofac/Autofac#988]
I have the following interfaces :
```
public interface ICommandHandler<T>
{
	void Handle(T command);
}
public class TransactionalCommandHandlerDecorator<T> : ICommandHandler<T>
{
    private readonly ICommandHandler<T> _handler;
    public TransactionalCommandHandlerDecorator(ICommandHandler<T> handler)
    {
        _handler = handler;
    }
    public void Handle(T command)
    {
    }
}
```
and I have a concrete class which implementing two command handlers :

```
public class Handler : ICommandHandler<CreateLocation>
                        ,ICommandHandler<ModifyLocation>
{
    public void Handle(CreateLocation command)
    {

    }

    public void Handle(ModifyLocation command)
    {
    }
}
```
and I have a registration as following :

```
 builder.RegisterAssemblyTypes(Assembly.GetExecutingAssembly())
            .AsClosedTypesOf(typeof(ICommandHandler<>))
            .InstancePerLifetimeScope();

    builder.RegisterGenericDecorator(typeof(TransactionalCommandHandlerDecorator<>), typeof(ICommandHandler<>));
```

Resolving 'Handler' class causes autofac to circularly resolving decorater and handler in an infinite loop which causes a StackOverflowException. if I a change 'Handler' to implement only one interface, then it will works without problem.

am I doing something wrong? or is it a bug?

**Comment by tillig** [comment]
[Cross posted.](https://stackoverflow.com/q/56153304/8116) Let's handle it there for now and if it turns out it's unsolvable we can reopen.

**Comment by alexmg** [comment]
This should be resolved with #1023.
